### ISSUE TITLE
WebP filter to generate image srcset for .webp files

### ISSUE THREAD

**Issue by HanifBirgani** [timber/timber#2285]
## Is your feature request related to a problem? Please describe.

I've liked to use the `towebp` filter on the `srcset` of the images, but the filter works only on a single file.

## Describe the solution you’d like

I wrote the function below to filter an image object and generate srcset of webp files:

```php
add_action('timber/twig/filters', function ($twig) {
    $twig->addFilter(
        new TwigFilter('towebpSrcset', function ($img) {
            $dir = dirname($img->guid);
            $arr = [];
            foreach ($img->sizes as $size) {
                array_push($arr, ImageHelper::img_to_webp($dir.'/'.$size['file']) .' '. $size['width'] .'w');
            }
            return join(', ', $arr);
        })
    );
    return $twig;
});
```

So I can use it on my twig files like this:
```twig
{{ imgObject|towebpSrcset }}
```
and it will generate the srcset of webp files for the given image object.

I like to put this feature in the Timber core. I would like to hear your opinions.

**Comment by jarednova** [comment]
@HanifBirgani I like where you're going, but interested to see if there's a more generic way of describing the utility? Perhaps if `towebp` returned an array when sent its argument as an array, one could use ...

`img_array | towebp | join(', ')`

... would that satisfy your use-case?

**Comment by gchtr** [comment]
Do I understand it right that the function takes the image sizes that were defined through `add_image_size()`? In that case, I think there should be a way to pass your own images sizes, because oftentimes, the default image sizes don’t match with what you need exactly.

```twig
<img srcset="{{ img.sizes|towebp|join(', ') }}">
<img srcset="{{ img|towebp(custom_sizes)|join(', ') }}">
```

And how would it work in combination with the `resize` filter?

**Comment by jarednova** [comment]
@gchtr I think we'd need to adapt the `resize` filter to accept arrays too. That would bring up a bunch of questions around how the arguments should be structured that would require a deep knowhow ...

```twig
<img srcset="{{ img.sizes | resize([800, 600],[1600, 1200],[2400, 1800]) | towebp | join(', ') }}">
```

... not impossible, but already I can feel my head spinning. 

There's currently an `Image::srcset` method. I feel like this is where users should be able to set something like:

```twig
<img srcset="{{ img.srcset({format: 'webp'}) }}">
```

... we'd need to do a little monkeying with the `Image::srcset` argument order, but same patterns we've used elsewhere to migrate to argument arrays.

**Comment by HanifBirgani** [comment]
I think one of the base problems is the way which Timber filters work. Currently, the `towebp` and `resize` filters operate on a single URL string, but the output of `Image::srcset` is not a URL string, because it directly returns the output of `wp_get_attachment_image_srcset`.

I suggest doing two things to maintain backward compatibility and solve the problems mentioned above:

1. Modifying the filters (or image helpers) to accept arrays as input too.
2. Defining a new property for the `Image` (like `Image::srcsetArray`) to return image sizes as an array `[ (url1, size1), (url2, size2), ... ]`

The twig code would be like:
```twig
<img srcset="{{ img.srcsetArray | towebp }}">
```

... then in the image helpers (e.g., `img_to_webp`), we should check if the input is a string or an array. If the input is a simple URL string, the filter will do its primary job like before. On the other hand, if the input is an array, the filter will loop through its items, do the operation on all URLs, then return a concatenated srcset string:

(`img_to_webp` pseudocode)
```
IF the input is a string
    convert the image to webp
    RETURN webp url

ELSE IF the input is an array
    DEFINE an empty srcset string
    FOR EACH image size in the input
         convert the image to webp
         APPEND webp url and size to the srcset string
    END FOR
    RETURN srcset string
END IF
```

The `resize` filter needs more details to be backward compatible:

```twig
<img srcset="{{ img.srcsetArray | resize(800, 600) }}">
```

(`resize` helper pseudocode)
```
IF the input is a string
    resize the image to (width, height)
    RETURN resized url

ELSE IF the input is an array
    IF (width, height) is in the input array
        RETURN input array
    ELSE
        resize the image to (width, height)
        APPEND resized url and size to the input array
    END IF
    create srcset string by joining the array
    RETURN srcset string
END IF
```

It would be more complicated if the user wants to pass multiple sizes to resize filter, and it's not currently supported by timber.

**Comment by gchtr** [comment]
> I think one of the base problems is the way which Timber filters work.

It definitely is.

I want to throw in a couple of links to issues here, because we already thought about different ways to handle image filters and use cases before, and the bigger picture should definitely be considered. (Everything is a bit hidden in a couple of places, sorry for that).

- https://github.com/timber/timber/issues/857#issuecomment-364782702 – Definitely read this one
- https://github.com/timber/timber/pull/1916
- https://github.com/timber/timber/pull/1942

Currently I don’t know what the best way to go here is. But my hunch is still that we should maybe have only one filter for image manipulation that works with an array of input arguments and ditch the rest of the filters, including `|webp`. In that sense, the thing that @jarednova proposed feels like a good way to solve it on the first look.

> ```twig
> <img srcset="{{ img.srcset({format: 'webp'}) }}">
> ```

But that doesn’t speak against looking at other solutions.

**Comment by HanifBirgani** [comment]
>... filter for image manipulation that works with an array of input arguments ...

I agree with you about input arrays completely. What about backward compatibility? Do you wanna drop support for themes that use old filters?
IMO dropping support is a reasonable move regarding the flexibility of input arrays.

**Comment by jarednova** [comment]
@HanifBirgani @gchtr the uses of images has changed a lot since those filters were originally developed. I'm all for taking a fresh crack at best-practices here to make sure we're considering the state of media handling in 2020. However, I feel like this is something we should defer as a team until we have 2.x out the door and then can triage a big change like this against other needs

**Comment by HanifBirgani** [comment]
I think [Mozilla responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images) could be a good starting point to decide which features should be implemented in the filter.
